q2_longitudinal/_longitudinal.py

```diff
@@ -128,12 +129,48 @@ def pairwise_distances(output_dir: str, distance_matrix: skbio.DistanceMatrix,
 
 
 def linear_mixed_effects(output_dir: str, metadata: qiime2.Metadata,
-                         metric: str, state_column: str,
-                         individual_id_column: str, group_columns: str = None,
+                         state_column: str, individual_id_column: str,
+                         metric: str = None, group_columns: str = None,
                          random_effects: str = None,
                          table: pd.DataFrame = None,
                          palette: str = 'Set1', lowess: bool = False,
-                         ci: int = 95) -> None:
+                         ci: int = 95, formula: str = None) -> None:
+
+    # Must use formula and/or metric.
+    if metric is None and formula is None:
+        raise ValueError('Must specify either a metric or a formula that '
+                         'contains a valid metric to use as a dependent '
+                         'variable.')
+
+    # Formulae must contain state_column and designate a metric.
+    # (state_column is used separately for plotting but is ambiguous in the
+    # formula, so it does make sense to require even when formula is passed.)
+    if formula is not None:
+        if '~' not in formula:
+            raise ValueError('Formula must be in format "metric ~ independent '
+                             'variables".')
+        if state_column not in formula:
+            raise ValueError(
+                '"formula" must contain the "state_column" value: '
+                '{0} does not contain {1}'.format(formula, state_column))
+
+    # optionally parse R-style formula for validation. Note that this will
+    # override "metric" and "group_columns" parameters if input separately.
+    # "formula" is meant to be a "secret" feature for power users familiar with
+    # R-style formulae, so I think it is okay to just clarify this in the docs
+    # instead of putting in too many safety features (e.g., to prevent this
+    # override behavior).
+    if formula is not None:
```

Comment:
I high-level comment here @nbokulich: can we use the built-in patsy parsing, instead of rolling our own? https://patsy.readthedocs.io/en/latest/formulas.html